src/btree/bt_compact.c

```diff
@@ -65,6 +65,34 @@ __compact_rewrite(WT_SESSION_IMPL *session, WT_REF *ref, int *skipp)
 }
 
 /*
+ * __compact_skip --
+ *	Return if it may be worthwhile compacting a file.
+ */
+static int
+__compact_skip(WT_SESSION_IMPL *session, int *skipp)
+{
+	WT_BTREE *btree;
+	WT_PAGE_INDEX *pindex;
+
+	*skipp = 1;
+
+	btree = S2BT(session);
+
+	/*
+	 * It's safe to look at the pindex of the root page here. Trees with
+	 * less than 4 level one pages aren't worth compacting.
+	 */
+	pindex = WT_INTL_INDEX_GET_SAFE(btree->root.page);
+	if (pindex->entries < 4)
```

Comment:
My initial thoughts for adding a check in the btree layer was looking at how many checkpoints exist in a file. The current check for whether subsequent checkpoints for compact are making progress is very naive (if a single page is touched by a checkpoint, we continue doing more checkpoints). What I was hoping for by adding a layer here was moving towards being able to track how much space subsequent checkpoints are reclaiming.

Now that I write that down, maybe it doesn't belong here either. I'll make the changes you suggested and open a new issue to enhance compact termination conditions.